crates/pallets/starknet/src/lib.rs

```diff
@@ -1092,64 +1090,82 @@ impl<T: Config> Pallet<T> {
+                                &mut BlockifierStateAdapter::<T>::default(),
+                                block_context,
+                                true,
+                                disable_nonce_validation,
+                            )
+                        }
+                    };
+                    execution_results.push(result);
+                }
                 storage::TransactionOutcome::Rollback(Ok(()))
             });
-            execution_result
+            execution_results
         }
 
-        let mut blockifier_state_adapter = BlockifierStateAdapter::<T>::default();
-        let block_context = Self::get_block_context();
-        let disable_nonce_validation = T::DisableNonceValidation::get();
-
-        let execution_result = match transaction {
-            UserTransaction::Declare(tx, contract_class) => execute_tx_and_rollback(
-                tx.try_into_executable::<T::SystemHash>(chain_id, contract_class, is_query)
-                    .map_err(|_| Error::<T>::InvalidContractClass)?,
-                &mut blockifier_state_adapter,
-                &block_context,
-                disable_nonce_validation,
-            ),
-            UserTransaction::DeployAccount(tx) => execute_tx_and_rollback(
-                tx.into_executable::<T::SystemHash>(chain_id, is_query),
-                &mut blockifier_state_adapter,
-                &block_context,
-                disable_nonce_validation,
-            ),
-            UserTransaction::Invoke(tx) => execute_tx_and_rollback(
-                tx.into_executable::<T::SystemHash>(chain_id, is_query),
-                &mut blockifier_state_adapter,
-                &block_context,
-                disable_nonce_validation,
-            ),
-        };
+        let execution_results = execute_txs_and_rollback::<T>(
+            transactions,
+            &Self::get_block_context(),
+            T::DisableNonceValidation::get(),
+            chain_id,
+        );
 
-        match execution_result {
-            Ok(tx_exec_info) => {
-                log!(debug, "Successfully estimated fee: {:?}", tx_exec_info);
-                if let Some(gas_usage) = tx_exec_info.actual_resources.0.get("l1_gas_usage") {
-                    Ok((tx_exec_info.actual_fee.0 as u64, *gas_usage as u64))
-                } else {
-                    Err(Error::<T>::TransactionExecutionFailed.into())
+        let mut results = vec![];
+        for res in execution_results {
+            match res {
+                Ok(tx_exec_info) => {
+                    log!(info, "Successfully estimated fee: {:?}", tx_exec_info);
+                    if let Some(l1_gas_usage) = tx_exec_info.actual_resources.0.get("l1_gas_usage") {
+                        results.push((tx_exec_info.actual_fee.0 as u64, *l1_gas_usage as u64));
+                    } else {
+                        return Err(Error::<T>::TransactionExecutionFailed.into());
+                    }
+                }
```

Comment:
This is not needed anymore if you stop executing the tx after the first that failed